File: platforms/Leetcode/DailyChallenge/March/week4/Vowel_Spellchecker.cpp
class Solution {
public:
    vector<string> spellchecker(vector<string>& wordlist, vector<string>& queries) {
    unordered_set<string> words;
    unordered_map<string, string> lower_case, no_vowel;

    for (auto word : wordlist) {
            words.insert(word);
            lower_case.insert({to_lower(word), word});
            no_vowel.insert({de_vowels(word), word});
        }
       vector<string> res;

        for (auto query : queries) {
            string lowercase = to_lower(query), novowels = de_vowels(query);

            if (words.find(query) != words.end())
                res.push_back(query);

            else if (lower_case.find(lowercase) != lower_case.end())
                res.push_back(lower_case[lowercase]);

            else if (no_vowel.find(novowels) != no_vowel.end()){
                res.push_back(no_vowel[novowels]);
            }

            else res.push_back("");
        }
        return res;
    }


    private:
    string to_lower(string word) {
       transform(word.begin(), word.end(), word.begin(), ::tolower);
        return word;
    }
    //remove vowels
    string de_vowels(string word) {
        unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u'};
        word=to_lower(word);
        for (auto& letter : word) {
            if (vowels.find(letter) != vowels.end()) {
                letter = '*';
            }
        }
        return word;
    }
};

File: platforms/Leetcode/DailyChallenge/March/week4/Vowel_Spellchecker.py
class Solution:

    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        words = set(wordlist)
        lower_case = defaultdict(list)
        no_vowel = defaultdict(list)
        vowels = set(['a', 'e', 'i', 'o', 'u'])
        res = []

        def de_vowels(word):
            no_vowels = []
            for ch in word:
                ch = ch.lower()
                if ch in vowels: no_vowels.append('*')
                else: no_vowels.append(ch)
            return ''.join(no_vowels)

        for word in wordlist:
            lower_case[word.lower()].append(word)
            no_vowel[de_vowels(word)].append(word)

        for query in queries:
            if query in words:
                res.append(query)
            elif (lowercase := query.lower()) in lower_case:
                res.append(lower_case[lowercase][0])
            elif (novowels := de_vowels(query)) in no_vowel:
                res.append(no_vowel[novowels][0])
            else:
                res.append('')

        return res
